### PR TITLE
Refactor reference resolving

### PR DESCRIPTION
**Pull request by wouterj**
Pfew, this has become larger than I thought.

This PR centralizes reference resolving and improves the link parsing phase. The two main changes:

* **Introduce centralized reference resolving.** At the start of the render phase, dedicated reference resolvers now make sure all link nodes have a URL set. This centralizes the logic which previously lived in compiling phase (through document transformers) and templates (through Twig functions). We need to use the render phase, as we need the `RenderContext` to figure out the correct file extension to use. Later on, we can use the render context to also generate relative URLs.
  I've done this in a decorating node rendering, as I didn't want to traverse the whole tree yet another time in the process. It's a bit less clean though, so let me know how you think about this.
* **Parser rules are free of 'reference details'.** Meaning they now parse `` `label <embedded URI>`_`` and no longer the Intersphinx implementation details of this `embedded URI`. This is left to the reference resolver.
  We didn't seem to do intersphinx resolving at the moment, I've written a resolver to showcase the functionality but this probably needs fixing in another PR.

---

@linawolf I used your branch as a base, but as these changes almost never overlap, I decided to rebase onto the main branch to ease review.

### CODE REVIEW

**Comment by linawolf** [comment]
@wouterj tests are red, is this still a draft?

**Comment by wouterj** [comment]
Oh I forgot about that one. Somehow an internal target to a section no longer has a title (while it has in other tests). I'll need to debug that one a bit more

**Comment by wouterj** [comment]
Apart from that one failure, the PR should be ready for review.

**Review by jaapio — COMMENTED** [review]
I think the concept you reintroduced for the reference resolvers is a good idea to have back. I found it complex to create my own type of reference. So thanks for that. 

Regarding the decorator it feels like this is not the cleanest solution. What would be the impact when people want to add their own decorator around the DelegatingRenderer? Do they need to decorate yours or is that handled correctly? I don't know what symfony DI will do in this case? Does it allow you to add multiple decorators? 

A possible other solution would be a new node renderer for the LinkInlineNode? As I think this solved the issue?

File: packages/guides-restructured-text/src/RestructuredText/Parser/InlineLexer.php [hunk 1]
```diff
@@ -40,6 +43,8 @@ final class InlineLexer extends AbstractLexer
...
...ings-privacy|ms-settings-proximity|ms-settings-screenrotation|ms-settings-wifi|ms-settings-workplace|ms-spd|ms-stickers|ms-sttoverlay|ms-transit-to|ms-useractivityset|ms-virtualtouchpad|ms-visio|ms-walk-to|ms-whiteboard|ms-whiteboard-cmd|ms-word|msnim|msrp|msrps|mss|mt|mtqp|mumble|mupdate|mvn|news|nfs|ni|nih|nntp|notes|num|ocf|oid|onenote|onenote-cmd|opaquelocktoken|openpgp4fpr|otpauth|p1|pack|palm|paparazzi|payment|payto|pkcs11|platform|pop|pres|prospero|proxy|pwid|psyc|pttp|qb|query|quic-transport|redis|rediss|reload|res|resource|rmi|rsync|rtmfp|rtmp|rtsp|rtsps|rtspu|sarif|secondlife|secret-token|service|session|sftp|sgn|shc|shttp (OBSOLETE)|sieve|simpleledger|simplex|sip|sips|skype|smb|smp|sms|smtp|snews|snmp|soap\.beep|soap\.beeps|soldat|spiffe|spotify|ssb|ssh|starknet|steam|stun|stuns|submit|svn|swh|swid|swidpath|tag|taler|teamspeak|tel|teliaeid|telnet|tftp|things|thismessage|tip|tn3270|tool|turn|turns|tv|udp|unreal|upt|urn|ut2004|uuid-in-package|v-event|vemmi|ventrilo|ves|videotex|vnc|view-source|vscode|vscode-insiders|vsls|w3|wais|web3|wcr|webcal|web+ap|wifi|wpid|ws|wss|wtai|wyciwyg|xcon|xcon-userid|xfire|xmlrpc\.beep|xmlrpc\.beeps|xmpp|xri|ymsgr|z39\.50|z39\.50r|z39\.50s)';
```
[the remark this answers is not in the document]
Comment (linawolf) [reply packages/guides-restructured-text/src/RestructuredText/Parser/InlineLexer.php:46]: should we really support schemas beside http and https? You can always make an explicit link if you want to link to other schemas. But I believe implicit link detection should not cover very complex cases

[the remark this answers is not in the document]
Comment (wouterj) [reply packages/guides-restructured-text/src/RestructuredText/Parser/InlineLexer.php:46]: This is following [the specs](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#standalone-hyperlinks):

> The scheme is the name of the protocol, such as "http", "ftp", "mailto", or "telnet". The scheme consists of an initial letter, followed by letters, numbers, and/or "+", "-", ".". Recognition is limited to known schemes, per the [Official IANA Registry of URI Schemes](http://www.iana.org/assignments/uri-schemes) and the W3C's [Retired Index of WWW Addressing Schemes](https://www.w3.org/Addressing/schemes.html).
> [...] A reStructuredText parser must be able to recognize any absolute URI, as defined in [RFC2396](https://www.rfc-editor.org/rfc/rfc2396.txt) and [RFC2732](https://www.rfc-editor.org/rfc/rfc2732.txt).

We also have a test case testing support for `news:` and `ftp://`.

**Comment by wouterj** [comment]
> Regarding the decorator it feels like this is not the cleanest solution. What would be the impact when people want to add their own decorator around the DelegatingRenderer? Do they need to decorate yours or is that handled correctly? I don't know what symfony DI will do in this case? Does it allow you to add multiple decorators?

You'll always need to decorate the main service (`NodeRendererFactory`), Symfony will stack the decorators correctly. Using [decoration priority](https://symfony.com/doc/current/service_container/service_decoration.html#decoration-priority), you can also change whether it'll be `DecoratorA(DecoratorB(NodeRendererFactory))` or `DecoratorB(DecoratorA(NodeRendererFactory))`.

But I agree that this solution does not seem very clean.

> A possible other solution would be a new node renderer for the LinkInlineNode? As I think this solved the issue?

The problem is knowing what the "real" link node renderer should be. Link resolving is format-independent, while the real node renderer is format dependent. Also, I'm not sure if this works for node interfaces, or only for concrete node classes?

Maybe the best solution is to add an event `PreRenderNode` or the like, and do the link resolving in a listener? The only issue is finding the best place for this event to be dispatched.

**Comment by jaapio** [comment]
Maybe we should introduce an event decorator? Like you did with the decorator you created right now. And introduce a listener for the reference resolving? This would enable a more generic hooking point for extensions. And we don't have to worry about when to dispatch events.

**Comment by jaapio** [comment]
Thanks @wouterj I finished your proposal in #480

**Pull request closed without merging**